Add quality templates from products into productions.

juanjo-nan opened the pull request:
Add cron to create quality test in productions.
Fix views.
Dont end a production until all the quality test are successful.

043843

juanjo-nan committed:
Add quality templates from products into productions.
Add cron to create quality test in productions.
Fix views.
Dont end a production until all the quality test are successful.

043843

raimonesteve commented on production.py:
```diff
@@ -1,38 +1,114 @@
 class Template(metaclass=PoolMeta):
     __name__ = 'product.template'
-    quality_template = fields.Many2One('quality.template', 'Quality Template')
+    quality_templates = fields.One2Many('product.template-quality.template', 'template', "Quality Templates")
```
Si canviem de m2o a o2m, que passa amb teb i altres entorns que tinguin aquest mòdul? hi ha script per "migrar" les dades?

juanjo-nan replied:
És un mòdul que només està teb, si es fa el canvi ara, no afectarà cap altre client: [ho pots veure als components](https://nan.nan-tic.com:8000/#nan/model/project.work.component/339;name=%22Components%22&views=%5B459%2C460%5D)

raimonesteve commented on production.py:
```diff
@@ -1,38 +1,114 @@
+    time_since_quality_control = fields.DateTime('Time since quality control')
 
-    @fields.depends('product')
-    def on_change_product(self):
```
Es millor canviar el m2p per un function(o2m) i no relacionar-ho amb el on_change? No seria millor "desar" les plantilles actuals a la producció, en contres del function, ja que que cada cop que es canvi un plantilla en el producte, quedarà "canviat" a totes les produccions?

juanjo-nan replied:
Ho he parlat amb l'Albert, en aquest cas s'ha decidit que és millor deixar-lo així.

raimonesteve commented on production.py:
```diff
@@ -1,38 +1,114 @@
+        quality_templates = []
+        if self.product:
+            for quality_template in self.product.template.quality_templates:
+                if quality_template.company.id == context.get('company'):
```
En principi aquesta condició no caldria pq "product.template.quality_templates" t'hauria de retornar les de teva empresa en que estàs si es crea al regla d'accés. (a no ser que es vulgui ja preparar pel drop company rules)

juanjo-nan replied:
Amb el canvi que es farà la 6.0 amb les companyies, es millor deixar-lo així.

raimonesteve commented on production.py:
```diff
@@ -1,38 +1,114 @@
+    def run(cls, productions):
+        for production in productions:
+            production.time_since_quality_control = datetime.now()
+            production.save()
```
Personalment m'agrada més un sol write en bloc.

```
cls.write(to_write)
```

juanjo-nan replied:
Ho he parlat amb l'Albert, al final seria millor fer un save massiu de totes les produccions afectades.

raimonesteve commented on production.py:
```diff
@@ -1,38 +1,114 @@
+            document = 'production,%s' % str(production.id)
+            tests_not_successful = QualityTest.search([
+                ('state', 'not like', 'successful'),
+                ('document', '=', 'production,%s' % str(production.id)),
```
Ja que defineixes la variable, es pot fer:

```
('document', '=', document),
```

raimonesteve commented on production.py:
```diff
@@ -1,38 +1,114 @@
+            if len(tests_not_successful) != 0:
+                raise UserError(gettext('production_quality_control.'
+                        'msg_test_not_successful',
+                        production=production.number))
```
Sóc més partidari per obtenir el nom descriptiu, fer:

```
production.rec_name
```

juanjo-nan committed:
Use Queue instead of cron.
Pyflakes fixes.

043843

juanjo-nan committed:
Calculate how many test to crete between cron executions.

043843

juanjo-nan committed:
Delete quality_templates from view.

juanjo-nan committed:
Add action to see quality test from the productions.

albertca commented on production.py:
```diff
@@ -1,38 +1,126 @@
+                ('state', 'not like', 'successful'),
+                ('document', '=', 'production,%s' % str(production.id)),
+            ])
+            if len(tests_not_successful) != 0:
```
D'això n'hem de parlar perquè si no admetem controls de qualitat fallits, llavors no es podrà tancar mai la producció en qüestió i això dubto que sigui un comportament adequat.

albertca commented on production.py:
```diff
@@ -1,38 +1,126 @@
+    "Production Template"
+    __name__ = 'product.template-quality.template'
+
+    template = fields.Many2One('product.template', "Template")
```
Falta required=True i ondelete='CASCADE'

albertca commented on production.py:
```diff
@@ -1,38 +1,126 @@
+    company = fields.Many2One('company.company', "Company", required=True)
+    interval = fields.Integer("Interval", required=True,
+        help="Interval in minutes")
+    quality_template = fields.Many2One('quality.template', "Quality Template")
```
Falta required=True

albertca commented on production.py:
```diff
@@ -1,38 +1,126 @@
             },
-        depends=['quality_template'])
+        depends=['quality_templates'])
+    time_since_quality_control = fields.DateTime('Time since quality control')
```
Crec que l'hauríem de fer readonly

juanjo-nan committed:
Fix cron method.
Add missing required in fields.
Delete unnecessary done method (we dont need to check if the production
have some tets failed).

albertca commented on production.py:
```diff
@@ -1,38 +1,114 @@
+                        to_save.append(test)
+
+                    production.time_since_quality_control = datetime.now()
+                    production.save()
```
You should save all productions at once at the end of the loop:

cls.save(productions)

juanjo-nan committed:
Use cls.save() instead of use a save for each record.

juanjo-nan committed:
Add locales.
